packages/builder/builder-shared/src/patch.ts

```diff
@@ -0,0 +1,48 @@
+import assert from 'assert';
+import path from 'path';
+import { pathToFileURL } from 'url';
+import type {
+  Compiler as WebpackCompiler,
+  MultiCompiler as WebpackMultiCompiler,
+} from 'webpack';
+import type {
+  Compiler as RspackCompiler,
+  MultiCompiler as RspackMultiCompiler,
+} from '@rspack/core';
+
+const GLOBAL_PATCHED_SYMBOL: unique symbol = Symbol('GLOBAL_PATCHED_SYMBOL');
+
+declare global {
+  interface Location {
+    [GLOBAL_PATCHED_SYMBOL]?: true;
+  }
+}
+
+/** fix issue about dart2js: https://github.com/dart-lang/sdk/issues/27979 */
+export function patchGlobalLocation() {
+  const href = pathToFileURL(process.cwd()).href + path.sep;
+  const location = Object.freeze({ [GLOBAL_PATCHED_SYMBOL]: true, href });
+  global.location ||= location as unknown as Location;
+  assert(typeof global.location.href === 'string');
+}
```

Comment:
We can only create href and location variables if global.location is not defined:

```ts
export function patchGlobalLocation() {
  if (!global.location) {
    const href = pathToFileURL(process.cwd()).href + path.sep;
    const location = Object.freeze({ [GLOBAL_PATCHED_SYMBOL]: true, href });
    global.location = location as unknown as Location;
  }
}
```

Reply:
done.